python-package/xgboost/core.py

```diff
@@ -196,20 +200,8 @@ def __init__(self, data, label=None, missing=0.0,
         if weight is not None:
             self.set_weight(weight)
 
-        # validate feature name
-        if not feature_names is None:
-            if not isinstance(feature_names, list):
-                feature_names = list(feature_names)
-            if len(feature_names) != len(set(feature_names)):
-                raise ValueError('feature_names must be unique')
-            if len(feature_names) != self.num_col():
-                msg = 'feature_names must have the same length as data'
-                raise ValueError(msg)
-            # prohibit to use symbols may affect to parse. e.g. ``[]=.``
-            if not all(isinstance(f, STRING_TYPES) and f.isalnum()
-                       for f in feature_names):
-                raise ValueError('all feature_names must be alphanumerics')
         self.feature_names = feature_names
+        self.feature_types = feature_types
```

Comment:
This is intentional to use setter which validates user input.

https://github.com/dmlc/xgboost/pull/503/files#diff-fd53d68e0037d3512896122d1248d969R503